File: create-project-zip.js
import fs from 'fs';
import path from 'path';
import { execSync } from 'child_process';

console.log('Creating project zip file...');

// Create a temporary directory for the clean project
const tempDir = '/tmp/project-export';
const zipPath = './project-complete.zip';

try {
  // Remove existing temp directory if it exists
  if (fs.existsSync(tempDir)) {
    execSync(`rm -rf ${tempDir}`);
  }

  // Create temp directory
  fs.mkdirSync(tempDir, { recursive: true });

  // Copy project files, excluding unnecessary items
  const excludePatterns = [
    'node_modules',
    '.git',
    'dist',
    'build',
    '*.log',
    '.env*',
    'project-complete.zip',
    'create-project-zip.js',
    'apply-schema-fixes.js',
    'check-and-complete-questions.js',
    'create-vendor-responses.js',
    'execute-schema-fix.js',
    'fix-order-status.js',
    'house-relocation-questions-script.js',
    'setup-*.js',
    'user-manual',
    '*-backup.ts',
    '*-broken*.ts'
  ];

  // Copy files manually, excluding unwanted items
  const copyDirectory = (src, dest, excludePatterns) => {
    if (!fs.existsSync(dest)) {
      fs.mkdirSync(dest, { recursive: true });
    }

    const items = fs.readdirSync(src);
    
    for (const item of items) {
      const srcPath = path.join(src, item);
      const destPath = path.join(dest, item);
      
      // Check if item should be excluded
      const shouldExclude = excludePatterns.some(pattern => {
        if (pattern.startsWith('*') && pattern.endsWith('*')) {
          const middle = pattern.slice(1, -1);
          return item.includes(middle);
        } else if (pattern.startsWith('*')) {
          return item.endsWith(pattern.slice(1));
        } else if (pattern.endsWith('*')) {
          return item.startsWith(pattern.slice(0, -1));
        } else {
          return item === pattern;
        }
      });
      
      if (shouldExclude) continue;
      
      const stat = fs.statSync(srcPath);
      
      if (stat.isDirectory()) {
        copyDirectory(srcPath, destPath, excludePatterns);
      } else {
        fs.copyFileSync(srcPath, destPath);
      }
    }
  };

  copyDirectory('./', tempDir, excludePatterns);

  // Create a clean package.json without development-specific scripts
  const packageJsonPath = path.join(tempDir, 'package.json');
  if (fs.existsSync(packageJsonPath)) {
    const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));
    
    // Clean up scripts to be more generic
    packageJson.scripts = {
      "dev": "NODE_ENV=development tsx server/index.ts",
      "build": "tsc && vite build",
      "start": "node dist/server/index.js",
      "db:push": "drizzle-kit push",
      "db:studio": "drizzle-kit studio"
    };

    // Remove any Replit-specific dependencies if they exist
    if (packageJson.devDependencies) {
      delete packageJson.devDependencies['@replit/vite-plugin-cartographer'];
      delete packageJson.devDependencies['@replit/vite-plugin-runtime-error-modal'];
    }

    fs.writeFileSync(packageJsonPath, JSON.stringify(packageJson, null, 2));
  }

  // Create a README for setup instructions
  const readmePath = path.join(tempDir, 'README-SETUP.md');
  const readmeContent = `# Project Setup Instructions

## Prerequisites
- Node.js 18+ installed
- PostgreSQL database (local or cloud)

## Setup Steps

1. **Install dependencies:**
   \`\`\`bash
   npm install
   \`\`\`

2. **Database Setup:**
   - Create a PostgreSQL database
   - Set the DATABASE_URL environment variable:
     \`\`\`bash
     export DATABASE_URL="postgresql://username:password@localhost:5432/database_name"
     \`\`\`
   - Push the database schema:
     \`\`\`bash
     npm run db:push
     \`\`\`

3. **Environment Variables:**
   Create a \`.env\` file in the root directory with:
   \`\`\`
   DATABASE_URL=your_postgresql_connection_string
   NODE_ENV=development
   \`\`\`

4. **Run the application:**
   \`\`\`bash
   npm run dev
   \`\`\`

5. **Access the application:**
   Open http://localhost:5000 in your browser

## Project Structure
- \`client/\` - Frontend React application
- \`server/\` - Backend Express server
- \`shared/\` - Shared schema and types
- \`drizzle.config.ts\` - Database configuration

## Database Management
- View database: \`npm run db:studio\`
- Update schema: \`npm run db:push\`

## Notes
- The application uses Supabase for database operations in the current setup
- Modify the database connection in \`server/storage.ts\` if needed
- The frontend is built with React, TypeScript, and Tailwind CSS
- The backend uses Express with TypeScript
`;

  fs.writeFileSync(readmePath, readmeContent);

  // Create the zip file
  console.log('Creating zip archive...');
  execSync(`cd /tmp && zip -r project-complete.zip project-export/`);
  
  // Move zip to current directory
  execSync(`mv /tmp/project-complete.zip ${zipPath}`);

  // Clean up temp directory
  execSync(`rm -rf ${tempDir}`);

  console.log(`Project zip created successfully: ${zipPath}`);
  console.log('File size:', execSync(`ls -lh ${zipPath} | awk '{print $5}'`).toString().trim());

} catch (error) {
  console.error('Error creating zip:', error.message);
  process.exit(1);
}

File: create_project_zip.py
#!/usr/bin/env python3
import os
import zipfile
import sys
from pathlib import Path

def should_exclude(file_path, exclude_patterns):
    """Check if a file/directory should be excluded based on patterns."""
    file_name = os.path.basename(file_path)
    
    for pattern in exclude_patterns:
        if pattern.startswith('*') and pattern.endswith('*'):
            # Pattern like *backup*
            middle = pattern[1:-1]
            if middle in file_name:
                return True
        elif pattern.startswith('*'):
            # Pattern like *.log
            if file_name.endswith(pattern[1:]):
                return True
        elif pattern.endswith('*'):
            # Pattern like setup-*
            if file_name.startswith(pattern[:-1]):
                return True
        else:
            # Exact match
            if file_name == pattern:
                return True
    
    return False

def create_project_zip():
    print("Creating project zip file using Python...")
    
    # Define what to exclude
    exclude_patterns = [
        'node_modules',
        '.git',
        'dist',
        'build',
        '*.log',
        '.env*',
        'project-complete.*',
        'create-project-zip.js',
        'create_project_zip.py',
        'apply-schema-fixes.js',
        'check-and-complete-questions.js',
        'create-vendor-responses.js',
        'execute-schema-fix.js',
        'fix-order-status.js',
        'house-relocation-questions-script.js',
        'setup-*.js',
        'user-manual',
        '*-backup.ts',
        '*-broken*.ts',
        '__pycache__',
        '*.pyc',
        '.cache',
        '.pythonlibs',
        '.uv',
        'pyproject.toml',
        'uv.lock',
        '*.sql'
    ]
    
    # Get current directory (project root)
    project_root = Path('.')
    zip_filename = 'project-complete.zip'
    
    try:
        with zipfile.ZipFile(zip_filename, 'w', zipfile.ZIP_DEFLATED) as zipf:
            # Walk through all files and directories
            for root, dirs, files in os.walk(project_root):
                # Filter out excluded directories
                dirs[:] = [d for d in dirs if not should_exclude(os.path.join(root, d), exclude_patterns)]
                
                for file in files:
                    file_path = os.path.join(root, file)
                    
                    # Skip excluded files
                    if should_exclude(file_path, exclude_patterns):
                        continue
                    
                    # Skip the zip file itself if it exists
                    if file == zip_filename:
                        continue
                    
                    # Calculate relative path for the zip
                    rel_path = os.path.relpath(file_path, project_root)
                    
                    # Add file to zip with timestamp handling
                    try:
                        zipf.write(file_path, rel_path)
                        print(f"Added: {rel_path}")
                    except ValueError as e:
                        if "ZIP does not support timestamps before 1980" in str(e):
                            # Create ZipInfo manually with a valid timestamp
                            info = zipfile.ZipInfo(rel_path)
                            info.date_time = (1980, 1, 1, 0, 0, 0)  # Set to 1980-01-01
                            with open(file_path, 'rb') as src:
                                zipf.writestr(info, src.read())
                            print(f"Added (fixed timestamp): {rel_path}")
                        else:
                            raise
        
        # Get file size
        zip_size = os.path.getsize(zip_filename)
        zip_size_mb = zip_size / (1024 * 1024)
        
        print(f"\nProject zip created successfully: {zip_filename}")
        print(f"File size: {zip_size_mb:.2f} MB")
        print(f"Total files: {len(zipf.namelist()) if 'zipf' in locals() else 'Unknown'}")
        
        return True
        
    except Exception as e:
        print(f"Error creating zip file: {str(e)}")
        return False

if __name__ == "__main__":
    success = create_project_zip()
    sys.exit(0 if success else 1)